File: libraries/ocliterator.go
package ocliterator

import "container/list"
import "ocl"
// import "fmt"

type OclIterator struct {
  position int
  elements *list.List
  currentElement *list.Element
}

func createOclIterator() *OclIterator {
  var res *OclIterator
  res = &OclIterator{}
  return res
}

func NewOclIterator_Sequence(col *list.List) *OclIterator {
  var result *OclIterator
  result = createOclIterator()
  result.elements = col
  if col.Len() > 0 { 
    result.position = 1
    result.currentElement = col.Front()
  } else { 
    result.position = 0
    result.currentElement = nil
  } 

  return result
}

func NewOclIterator_Set(col *list.List) *OclIterator {
  var result *OclIterator
  result = createOclIterator()
  result.elements = col
  if col.Len() > 0 { 
    result.position = 1
    result.currentElement = col.Front()
  } else { 
    result.position = 0
    result.currentElement = nil
  } 
  return result
}

func NewOclIterator_String(s string) *OclIterator {
  var result *OclIterator
  result = createOclIterator()
  col := ocl.Split(s,"[ \\n\\t]+")
  result.elements = col
  if col.Len() > 0 { 
    result.position = 1
    result.currentElement = col.Front()
  } else { 
    result.position = 0
    result.currentElement = nil
  } 
  return result
}

func NewOclIterator_String_String(s string, r string) *OclIterator {
  var result *OclIterator
  result = createOclIterator()
  col := ocl.Split(s,"[" + r + "]+")
  result.elements = col
  if col.Len() > 0 { 
    result.position = 1
    result.currentElement = col.Front()
  } else { 
    result.position = 0
    result.currentElement = nil
  } 
  return result
}

func (self *OclIterator) At(i int) interface{} { 
  return ocl.At(self.elements,i)
} 

func (self *OclIterator) HasNext() bool { 
  return 0 <= self.position && 
         self.position <= self.elements.Len()
}

func (self *OclIterator) HasPrevious() bool {
  return self.position >= 1 &&
         self.position <= self.elements.Len() + 1
}

func (self *OclIterator) NextIndex() int { 
  return self.position + 1
}

func (self *OclIterator) PreviousIndex() int {
  return self.position - 1
}

func (self *OclIterator) MoveForward() { 
  if self.position <= self.elements.Len() { 
    self.position = self.position + 1
  }

  if (self.currentElement != nil) {
    self.currentElement = self.currentElement.Next() 
  }
}

func (self *OclIterator) MoveBackward() {
  if self.position >= 1 {
    self.position = self.position - 1
  }

  if (self.currentElement != nil) {
    self.currentElement = self.currentElement.Prev() 
  }
}

func (self *OclIterator) MoveTo(n int) { 
  if 0 <= n && n <= self.elements.Len() + 1 { 
    self.position = n
    i := 0
    for x := self.elements.Front(); x != nil; x = x.Next() { 
      if n == i+1 {
        self.currentElement = x
        return
      } 
      i = i+1
    }
  }
}

func (self *OclIterator) GetCurrent() interface{} { 
  if self.currentElement != nil { 
    return self.currentElement.Value
  } 

  if 1 <= self.position && 
     self.position <= self.elements.Len() {
    return ocl.At(self.elements,self.position) 
  }
  return nil
}

func (self *OclIterator) getElement() *list.Element {
  i := 0
  for x := self.elements.Front(); x != nil; x = x.Next() { 
    if self.position == i+1 {
      return x
    } 
    i = i+1
  }
  return nil
}

func (self *OclIterator) Set(v interface{}) { 
  if self.currentElement != nil { 
    self.currentElement.Value = v
    return
  } 

  i := 0
  for x := self.elements.Front(); x != nil; x = x.Next() { 
    if self.position == i+1 {
      x.Value = v
      return
    } 
    i = i+1
  }
}

func (self *OclIterator) Remove() { 
  if self.currentElement != nil { 
    newCurrent := self.currentElement.Next()
    self.elements.Remove(self.currentElement)
    self.currentElement = newCurrent
    return
  } 

  x := self.getElement()
  if x != nil {
    self.elements.Remove(x)
  }
}

func (self *OclIterator) Insert(v interface{}) { 
  if self.currentElement != nil { 
    self.currentElement = self.elements.InsertBefore(v,self.currentElement)
    return
  }

  x := self.getElement()
  if x != nil {
    self.elements.InsertBefore(v,x)
  }
}
  

func (self *OclIterator) Next() interface{} {
  curr := self.GetCurrent()
  self.MoveForward()
  return curr 
} 

func (self *OclIterator) Previous() interface{} {
  curr := self.GetCurrent()
  self.MoveBackward()
  return curr
} 

func (self *OclIterator) Length() int { 
  return self.elements.Len()
} 

File: ocliterator.py

import ocl
import math
import re
import copy

from oclfile import *
from enum import Enum

def free(x):
  del x


def displayint(x):
  print(str(x))

def displaylong(x):
  print(str(x))

def displaydouble(x):
  print(str(x))

def displayboolean(x):
  print(str(x))

def displayString(x):
  print(x)

def displaySequence(x):
  print(x)

def displaySet(x):
  print(x)

def displayMap(x):
  print(x)


class OclIteratorResult : 
  def __init__(self):
    self.value = None
    self.done = True

  def newOclIteratorResult(v) : 
    res = OclIteratorResult()
    res.value = v
    if v == None : 
      res.done = True
    else : 
      res.done = False
    return res



class OclIterator :
  ocliterator_instances = []
  ocliterator_index = dict({})

  def __init__(self):
    self.position = 0
    self.markedPosition = 0
    self.elements = []
    self.columnNames = []
    self.generatorFunction = None
    OclIterator.ocliterator_instances.append(self)


  def getPosition(self) : 
    return self.position

  def hasNext(self) :
    result = False
    if self.position >= 0 and self.position < len(self.elements) :
      result = True
    else :
      result = False
    return result

  def isAfterLast(self) :
    result = False
    if self.position > len(self.elements) : 
      return True
    return result

  def isBeforeFirst(self) :
    result = False
    if self.position <= 0 : 
      return True
    return result

  def hasPrevious(self) :
    result = False
    if self.position > 1 and self.position <= len(self.elements) + 1 :
      result = True
    else :
      result = False
    return result

  def nextIndex(self) :
    result = 0
    result = self.position + 1
    return result

  def previousIndex(self) :
    result = 0
    result = self.position - 1
    return result

  def moveForward(self) :
    self.position = self.position + 1

  def moveBackward(self) :
    self.position = self.position - 1

  def moveTo(self, i) :
    self.position = i

  def setPosition(self, i) :
    self.position = i

  def markPosition(self) :
    self.markedPosition = self.position

  def movePosition(self, i) :
    self.position = i + self.position

  def moveToFirst(self) :
    self.position = 1

  def moveToLast(self) :
    self.position = len(self.elements)

  def moveToStart(self) :
    self.position = 0

  def moveToEnd(self) :
    self.position = len(self.elements) + 1

  def moveToMarkedPosition(self) :
    self.position = self.markedPosition

  def close(self) : 
    self.position = 0
    self.markedPosition = 0
    self.columnNames = []
    self.elements = []

  def newOclIterator_Sequence(sq) :
    result = None
    ot = createOclIterator()
    ot.elements = sq
    ot.position = 0
    result = ot
    return result

  def newOclIterator_Set(st) :
    result = None
    ot = createOclIterator()
    ot.elements = ocl.unionSequence([], ocl.sortSet(st))
    ot.position = 0
    result = ot
    return result

  def newOclIterator_String(ss) :
    result = None
    ot = createOclIterator()
    ot.elements = ocl.split(ss, "[ \n\t\r]+")
    ot.position = 0
    result = ot
    return result

  def newOclIterator_String_String(ss,seps) :
    result = None
    ot = createOclIterator()
    ot.elements = ocl.split(ss, "[" + seps + "]+")
    ot.position = 0
    result = ot
    return result

  def newOclIterator_Function(f) :
    result = None
    ot = createOclIterator()
    ot.elements = []
    ot.position = 0
    ot.generatorFunction = f
    result = ot
    return result

  def trySplit(self) : 
    firstpart = self.elements[0:self.position]
    self.elements = self.elements[self.position:]
    self.position = 0
    return OclIterator.newOclIterator_Sequence(firstpart)

  def getCurrent(self) :
    result = None
    if self.position <= len(self.elements) and self.position >= 1 : 
      result = (self.elements)[self.position - 1]
    return result

  def set(self, x) :
    self.elements[self.position -1] = x

  def insert(self, x) :
    self.elements.insert(self.position-1,x)

  def remove(self) :
    self.elements[self.position -1] = None
    del self.elements[self.position -1]

  def tryAdvance(self, f) : 
    if self.position + 1 <= len(self.elements) : 
      x = self.next()  
      f(x) 
      return True
    return False 

  def next(self) :
    result = None
    self.moveForward()
    return self.getCurrent()

  def nextResult(self) :
    result = None
    if self.generatorFunction == None : 
      self.moveForward()
      v = self.getCurrent()
      return OclIteratorResult.newOclIteratorResult(v)
    else : 
      r = self.generatorFunction(self.position)
      self.position = self.position + 1
      if self.position <= len(self.elements) : 
        self.set(r)
      else : 
        self.elements.append(r)
      return OclIteratorResult.newOclIteratorResult(r)

  def forEachRemaining(self, f) : 
    for x in self.elements[self.position:] :
      f(x) 

  def previous(self) :
    result = None
    self.moveBackward()
    return self.getCurrent()

  def length(self) : 
    return len(self.elements)

  def at(self,i) : 
    return self.elements[i-1]

  def getColumnCount(self) : 
    return len(self.columnNames)

  def getColumnName(self,i) : 
    return self.columnNames[i-1]

  def getCurrentFieldByIndex(self,i) : 
    fld = self.columnNames[i-1]
    curr = self.getCurrent()
    if curr != None : 
      return curr[fld]
    return None

  def setCurrentFieldByIndex(self,i,v) : 
    fld = self.columnNames[i-1]
    curr = self.getCurrent()
    if curr != None : 
      curr[fld] = v


  def killOclIterator(ocliterator_x) :
    ocliterator_instances = ocl.excludingSet(ocliterator_instances, ocliterator_x)
    free(ocliterator_x)


def createOclIterator():
  ocliterator = OclIterator()
  return ocliterator

def allInstances_OclIterator():
  return OclIterator.ocliterator_instances




